src/coreclr/vm/jithelpers.cpp

```diff
@@ -589,17 +589,21 @@ HCIMPLEND
 HCIMPL1_V(UINT64, JIT_Dbl2ULng, double val)
 {
     FCALL_CONTRACT;
-
-    const double two63  = 2147483648.0 * 4294967296.0;
-    UINT64 ret;
-    if (val < two63) {
-        ret = FastDbl2Lng(val);
+    UINT64 unsigned_bits = *(UINT64 *)&val;
+    int sign_bit = int(unsigned_bits >> 63);
+    if (sign_bit > 0 && val != 0)
+    {
+        return UINT64_MAX;
     }
-    else {
-        // subtract 0x8000000000000000, do the convert then add it back again
-        ret = FastDbl2Lng(val - two63) + I64(0x8000000000000000);
+    
+    int exponent = int((unsigned_bits >> 52) & 0x7ff) - 1023;
+    if (exponent > 63)
+    {
+        return UINT64_MAX;
     }
-    return ret;
+
+    return FastDbl2Lng(val);
+
```

Comment:
Can this not just be `return ((UINT64)val);` now?

The previous logic was working around (and notably doing so incorrectly) a bug in some native compilers which I don't believe exists anymore.

Reply:
It can. It was kept as is to keep it in parallel to the previous implementation. Removing it in next commit.

Reply:
Changed the implementation to default to MSVC casting.